### ISSUE TITLE
TimeIntegrator API update

### ISSUE THREAD

**Issue by tkarna** [thetisproject/thetis#68]
I think we could simplify the time integrator API a little:
1. Omit solution function from `TimeIntegrator.advance` call
   
   Currently `TimeIntegrator.advance` takes the solution function as an argument. This is redundant as in practice `TimeIntegrator` needs to store the solution function as a property in any case. It's also difficult to ensure meaningful operation in case the user happens to pass a different function to the `advance` method.
2. Omit `dt` from `TimeIntegrator.advance` + add `set_dt` method
   
   We normally use a constant time step so there's no need to pass the time step to every call. In case the user wants to change the time step, I would add `set_dt` method instead. Then one could make sure that the time step gets updated correctly. In cases where changing time step is not allowed (e.g. linear multistep methods) we'd raise an exception.

Here's a draft of the base classes

``` python
class TimeIntegratorBase(object):
    """Base class for all time integrator objects."""
    __metaclass__ = ABCMeta

    @abstractproperty
    def advance(self, t, update_forcings=None):
        """Advances equations for one time step."""
        pass

    @abstractproperty
    def initialize(self, *args, **kwargs):
        """Initializes the time integrator"""
        pass


class TimeIntegrator(TimeIntegratorBase):
    """Base class for all time integrator objects that solve a single equation"""
    def __init__(self, equation, solution, fields, dt, solver_parameters={}):
        super(TimeIntegrator, self).__init__()

        self.equation = equation
        self.solution = solution
        self.fields = fields
        self.dt_const = Constant(dt)

        # unique identifier for solver
        self.name = '-'.join([self.__class__.__name__,
                              self.equation.__class__.__name__])
        self.solver_parameters = {}
        self.solver_parameters.update(solver_parameters)

    def set_dt(self, dt):
        """Updates time step"""
        self.dt_const.assign(dt)
```

Here I'm separating `TimeIntegratorBase` and `TimeIntegrator` classes: All time integrators are derived from the former (e.g. 3D coupled time integrators) while all "normal" time integrators (that iterate a single equation) inherit the latter.

Here I haven't changed the signature for `update_forcings` function which is still passed to the `advance` method. There might be a better way of handling that too.

**Comment by stephankramer** [comment]
Yes, both things seem very sensible.

**Comment by tkarna** [comment]
Now implemented in #70
